File: lab 3/Calendar.cpp
#include "Calendar.h"
#include "RecurringEvent.h"
#include "SingleEvent.h"
#include <iostream>
using namespace std;

void Calendar::addEvent(Event* event) {
    events.push_back(event);
}

Event* Calendar::findShortestEvent() const {
    int startDay, startMonth, startYear, endDay, endMonth, endYear;
    cout << "Enter start date (day month year): ";
    cin >> startDay >> startMonth >> startYear;
    cout << "Enter end date (day month year): ";
    cin >> endDay >> endMonth >> endYear;

    vector<Event*> filteredEvents = filterEventsByDate(startDay, startMonth, startYear, endDay, endMonth, endYear);

    if (filteredEvents.empty()) return nullptr;

    Event* shortestEvent = filteredEvents[0];
    for (const auto& event : filteredEvents) {
        if (event->getDuration() < shortestEvent->getDuration()) {
            shortestEvent = event;
        }
    }
    return shortestEvent;
}


Event* Calendar::findLongestEvent() const {

    int startDay, startMonth, startYear, endDay, endMonth, endYear;
    cout << "Enter start date (day month year): ";
    cin >> startDay >> startMonth >> startYear;
    cout << "Enter end date (day month year): ";
    cin >> endDay >> endMonth >> endYear;

    vector<Event*> filteredEvents = filterEventsByDate(startDay, startMonth, startYear, endDay, endMonth, endYear);

    if (filteredEvents.empty()) return nullptr;

    Event* longestEvent = filteredEvents[0];
    for (const auto& event : filteredEvents) {
        if (event->getDuration() > longestEvent->getDuration()) {
            longestEvent = event;
        }
    }
    return longestEvent;
}

double Calendar::calculateSumDuration() const {


    int startDay, startMonth, startYear, endDay, endMonth, endYear;
    cout << "Enter start date (day month year): ";
    cin >> startDay >> startMonth >> startYear;
    cout << "Enter end date (day month year): ";
    cin >> endDay >> endMonth >> endYear;

    vector<Event*> filteredEvents = filterEventsByDate(startDay, startMonth, startYear, endDay, endMonth, endYear);

    double total = 0.0;

    for (const auto& filteredEvents : filteredEvents) {
        auto recurringEvent = dynamic_cast<RecurringEvent*>(filteredEvents);
        if (recurringEvent) {
            total += recurringEvent->getDuration() * recurringEvent->getRepeats();
        }
        else {
            total += filteredEvents->getDuration();
        }
    }

    return total;
}


int Calendar::averageRepeats() const {
    int totalRepeats = 0, count = 0;
    for (const auto& event : events) {
        auto recurringEvent = dynamic_cast<RecurringEvent*>(event);
        if (recurringEvent) {
            totalRepeats += recurringEvent->getRepeats();
            count++;
        }
    }
    return count == 0 ? 0 : totalRepeats / count;
}

void Calendar::displayAllEvents() const {
    for (const auto& event : events) {
        event->displayEvent();
    }
}

bool Calendar::saveToFile(const string& filename) {
    return fileManager.saveEventsToFile(events, filename);
}

bool Calendar::loadFromFile(const string& filename) {
    return fileManager.loadEventsFromFile(events, filename);
}

vector<Event*> Calendar::filterEventsByDate(int startDay, int startMonth, int startYear, int endDay, int endMonth, int endYear) const {
    vector<Event*> filteredEvents;

    for (const auto& event : events) {
        SingleEvent* singleEvent = dynamic_cast<SingleEvent*>(event);
        RecurringEvent* recurringEvent = dynamic_cast<RecurringEvent*>(event);

        if (singleEvent) {

            const int* date = singleEvent->getDate();
            int day = date[0];
            int month = date[1];
            int year = date[2];

            if ((year > startYear || (year == startYear && (month > startMonth || (month == startMonth && day >= startDay)))) &&
                (year < endYear || (year == endYear && (month < endMonth || (month == endMonth && day <= endDay))))) {
                filteredEvents.push_back(singleEvent);
            }
        }
        else if (recurringEvent) {

            const int* startDate = recurringEvent->getStartDate();
            const int* endDate = recurringEvent->getEndDate();

            int sDay = startDate[0], sMonth = startDate[1], sYear = startDate[2];
            int eDay = endDate[0], eMonth = endDate[1], eYear = endDate[2];

            bool overlaps = (sYear < endYear || (sYear == endYear && (sMonth < endMonth || (sMonth == endMonth && sDay <= endDay)))) &&
                (eYear > startYear || (eYear == startYear && (eMonth > startMonth || (eMonth == startMonth && eDay >= startDay))));

            if (overlaps) {
                filteredEvents.push_back(recurringEvent);
            }
        }
    }

    return filteredEvents;
}

File: python/Calendar.py
from FileManager import FileManager
from SingleEvent import SingleEvent
from ReccuringEvent import ReccuringEvent


class Calendar:
    def __init__(self):
        self.events = []

    def add_event(self, event):
        self.events.append(event)

    def find_shortest_event(self):
        start_day, start_month, start_year = map(int, input("Enter start date (day month year): ").split())
        end_day, end_month, end_year = map(int, input("Enter end date (day month year): ").split())

        filtered_events = self.filter_events_by_date(start_day, start_month, start_year, end_day, end_month, end_year)

        if not filtered_events:
            return None

        shortest_event = filtered_events[0]
        for event in filtered_events:
            if event.get_duration() < shortest_event.get_duration():
                shortest_event = event

        return shortest_event

    def find_longest_event(self):
        start_day, start_month, start_year = map(int, input("Enter start date (day month year): ").split())
        end_day, end_month, end_year = map(int, input("Enter end date (day month year): ").split())

        filtered_events = self.filter_events_by_date(start_day, start_month, start_year, end_day, end_month, end_year)

        if not filtered_events:
            return None

        longest_event = filtered_events[0]
        for event in filtered_events:
            if event.get_duration() > longest_event.get_duration():
                longest_event = event

        return longest_event

    def calculate_sum_duration(self):
        start_day, start_month, start_year = map(int, input("Enter start date (day month year): ").split())
        end_day, end_month, end_year = map(int, input("Enter end date (day month year): ").split())

        filtered_events = self.filter_events_by_date(start_day, start_month, start_year, end_day, end_month, end_year)

        total = 0.0
        for event in filtered_events:
            if isinstance(event, ReccuringEvent):
                total += event.get_duration() * event.get_repeats()
            else:
                total += event.get_duration()

        return total

    def average_repeats(self):
        total_repeats = 0
        count = 0

        for event in self.events:
            if isinstance(event, ReccuringEvent):
                total_repeats += event.get_repeats()
                count += 1

        return 0 if count == 0 else total_repeats // count

    def display_all_events(self):
        for event in self.events:
            event.display_event()

    def save_to_file(self, filename):
        return FileManager.save_events_to_file(self.events, filename)

    def load_from_file(self, filename):
        return FileManager.load_events_from_file(self.events, filename)

    def filter_events_by_date(self, start_day, start_month, start_year,
                              end_day, end_month, end_year):
        filtered_events = []

        for event in self.events:
            if isinstance(event, SingleEvent):
                day, month, year = event.get_date()

                if ((year > start_year or (year == start_year and (month > start_month or (month == start_month and day >= start_day)))) and
                    (year < end_year or (year == end_year and (month < end_month or (month == end_month and day <= end_day))))):
                    filtered_events.append(event)

            elif isinstance(event, ReccuringEvent):
                s_day, s_month, s_year = event.get_start_date()
                e_day, e_month, e_year = event.get_end_date()

                overlaps = ((s_year < end_year or (s_year == end_year and (s_month < end_month or (s_month == end_month and s_day <= end_day)))) and
                            (e_year > start_year or (e_year == start_year and (e_month > start_month or (e_month == start_month and e_day >= start_day)))))

                if overlaps:
                    filtered_events.append(event)

        return filtered_events
